File: src/main/java/com/example/ui/pages/heroku/ContextMenuPage.java
package com.example.ui.pages.herokuapp;

import com.example.base.TestContext;
import com.example.ui.pages.BasePage;
import com.microsoft.playwright.Locator;
import com.microsoft.playwright.Page;
import com.microsoft.playwright.options.MouseButton;
import lombok.extern.slf4j.Slf4j;

import java.util.concurrent.atomic.AtomicReference;

/**
 * Context Menu page ({@code /context_menu}): right-clicking the hot-spot
 * raises a JS alert.
 */
@Slf4j
public class ContextMenuPage extends BasePage {

    public ContextMenuPage(Page page, TestContext testContext) {
        super(page, testContext);
    }

    /** Right-click the hot-spot and return the resulting alert text. */
    public String rightClickHotSpot() {
        AtomicReference<String> alertText = new AtomicReference<>();
        page.onceDialog(dialog -> {
            alertText.set(dialog.message());
            dialog.dismiss();
        });
        page.locator("#hot-spot").click(
                new Locator.ClickOptions().setButton(MouseButton.RIGHT));
        log.info("Context-click alert text: '{}'", alertText.get());
        return alertText.get();
    }
}

File: typescript/src/pages/heroku/context-menu.page.ts
import { BasePage } from '../base.page';

/** Context Menu page (/context_menu): right-clicking the hot-spot raises a JS alert. */
export class ContextMenuPage extends BasePage {
  /** Right-click the hot-spot and return the resulting alert text. */
  async rightClickHotSpot(): Promise<string | null> {
    let alertText: string | null = null;
    this.page.once('dialog', async (dialog) => {
      alertText = dialog.message();
      await dialog.dismiss();
    });
    await this.page.locator('#hot-spot').click({ button: 'right' });
    return alertText;
  }
}
